interpreter/cling/lib/Interpreter/DefinitionShadower.cpp

```diff
@@ -0,0 +1,189 @@
+        continue;
+      if (isa<FunctionTemplateDecl>(Prev) && isa<FunctionTemplateDecl>(D)
+          && m_Sema->IsOverload(cast<FunctionTemplateDecl>(D)->getTemplatedDecl(),
+                                cast<FunctionTemplateDecl>(Prev)->getTemplatedDecl(),
+                                /*IsForUsingDecl=*/false))
+        continue;
+
+      hideDecl(Prev);
+
+      // For unscoped enumerations, also invalidate all enumerators.
+      if (EnumDecl *ED = dyn_cast<EnumDecl>(Prev)) {
+        if (!ED->isTransparentContext())
+          continue;
+        for (auto &J : ED->decls())
+          if (NamedDecl *ND = dyn_cast<NamedDecl>(J))
+            hideDecl(ND);
+      }
+    }
+
+    // Ignore any forward declaration issued after a definition. Fixes "error
+    // : reference to 'xxx' is ambiguous" in `class C {}; class C; C foo;`.
+    if (!Previous.empty() && !isDefinition(D))
+      D->setInvalidDecl();
+  }
+
+  void DefinitionShadower::invalidatePreviousDefinitions(FunctionDecl *D) const {
+    const CompilationOptions &CO = getTransaction()->getCompilationOpts();
+    if (utils::Analyze::IsWrapper(D)) {
+      if (!CO.DeclarationExtraction)
+        return;
+
+      // DeclExtractor shall move local declarations to the TU. Invalidate all
+      // previous definitions (that may clash) before it runs.  
+      auto CS = dyn_cast<CompoundStmt>(D->getBody());
+      for (auto &I : CS->body()) {
+        auto DS = dyn_cast<DeclStmt>(I);
+        if (!DS)
+          continue;
+        for (auto &J : DS->decls())
+          if (auto ND = dyn_cast<NamedDecl>(J))
+            invalidatePreviousDefinitions(ND);
+      }
+    } else
+      invalidatePreviousDefinitions(cast<NamedDecl>(D));
+  }
+
+  void DefinitionShadower::invalidatePreviousDefinitions(Decl *D) const {
+    if (auto TD = dyn_cast<TagDecl>(D))
+      invalidatePreviousDefinitions(TD);
+    else if (auto FD = dyn_cast<FunctionDecl>(D))
+      invalidatePreviousDefinitions(FD);
+    else if (auto TD = dyn_cast<TemplateDecl>(D))
+      invalidatePreviousDefinitions(TD);
+  }
+
+  ASTTransformer::Result DefinitionShadower::Transform(Decl* D) {
+    const CompilationOptions &CO = getTransaction()->getCompilationOpts();
+    // Global declarations whose origin is the Cling prompt are subject to be
+    // nested in a `__cling_N5' namespace.
+    if (!CO.EnableShadowing
+        || D->getLexicalDeclContext() != m_TU || D->isInvalidDecl()
+        || isa<UsingDirectiveDecl>(D)
+        // FIXME: NamespaceDecl requires additional processing (TBD)
+        || isa<NamespaceDecl>(D)
+        || (isa<FunctionDecl>(D) && cast<FunctionDecl>(D)->isTemplateInstantiation())
+        || !typedInClingPrompt(FullSourceLoc{D->getLocation(),
+                                             m_Context.getSourceManager()}))
+      return Result(D, true);
+
+    NamespaceDecl *NS = NamespaceDecl::Create(m_Context, m_TU, /*inline=*/true,
```

Comment:
Could we have one `NamespaceDecl` per transaction?

Reply:
Ouch!  Fixed; thanks for the suggestion!